ykllvmwrap/src/ykllvmwrap.cc

```diff
@@ -37,15 +56,113 @@ extern "C" char *__yk_llvmwrap_symbolizer_find_code_sym(
     return strdup(LineInfo->FunctionName.c_str());
 }
 
-/// Compiles an IRTrace to executable code in memory.
+// Load an LLVM module from an address.
+std::unique_ptr<Module> load_module(LLVMContext &Context, char *Ptr, size_t Len) {
+    auto Sf = StringRef(Ptr, Len);
+    auto Mb = MemoryBufferRef(Sf, "");
+    SMDiagnostic Error;
+    auto M = parseIR(Mb, Error, Context);
+    if (!M)
+        errx(EXIT_FAILURE, "Can't load module.");
+    return M;
+}
+
+// Compile a module in-memory and return a pointer to its function.
+extern "C" void *compile_module(Module *M) {
+    InitializeNativeTarget();
+    InitializeNativeTargetAsmPrinter();
+    InitializeNativeTargetAsmParser();
+
+    // FIXME Remember memman or allocated memory pointers so we can free the
+    // latter when we're done with the trace.
+    auto memman = new MemMan();
+
+    auto MPtr = std::unique_ptr<Module>(M);
+    ExecutionEngine *EE = EngineBuilder(std::move(MPtr))
+        .setMemoryManager(std::unique_ptr<MCJITMemoryManager>(memman))
+        .create();
+	EE->finalizeObject();
+
+    if (EE->hasError())
+        errx(EXIT_FAILURE, "Couldn't compile trace: %s", EE->getErrorMessage().c_str());
+
+    return (void *) EE->getFunctionAddress(TRACE_FUNC_NAME);
+}
+
+// Compiles an IRTrace to executable code in memory.
 //
 // The trace to compile is passed in as two arrays of length Len. Then each
 // (FuncName[I], BBs[I]) pair identifies the LLVM block at position `I` in the
 // trace.
-extern "C" void __ykllvmwrap_irtrace_compile(char *FuncNames[], size_t BBs[], size_t Len) {
+//
+// Returns a pointer to the compiled function.
+extern "C" void *__ykllvmwrap_irtrace_compile(char *FuncNames[], size_t BBs[], size_t Len, char *SecPtr, size_t SecSize) {
+    LLVMContext Context;
+    auto DstMod = new Module("", Context);
+
+    // Set up new module for the trace.
+    llvm::FunctionType *FType = llvm::FunctionType::get(Type::getVoidTy(Context), false);
+	llvm::Function *DstFunc = llvm::Function::Create(FType, Function::InternalLinkage, TRACE_FUNC_NAME, DstMod);
+    DstFunc->setCallingConv(CallingConv::C);
```

Comment:
I think this will need a fixme for matching the calling convention of the original function.

Reply:
This is our newly created function which contains our trace, and will be run by the meta-tracer. So I think we do want C or SysV here. I could be wrong though.